### PR TITLE
kokkos_tribits: check for libdl with Trilinos

### PR DESCRIPTION
**Pull request by ndellingwood**
Kokkos_ENABLE_LIBDL was not being set in Trilinos builds, this is a
workaround to allow for the check.
Reported by @DavidPoliakoff and @tasmith4

### CODE REVIEW

**Comment by ndellingwood** [comment]
This workaround allowed me to get Kokkos_ENABLE_LIBDL enabled in Trilinos, a better solution to supercede this PR should be part of the 3.2.1 patch release

**Review by masterleinad — APPROVED** [review]
I was looking into doing something similar but was failing on understanding how TPL integration works in Trilinos. If this works, it's totally fine with me. It seems that no other package is depending on LIBDL right now anyway.

**Comment by ndellingwood** [comment]
CI errors unrelated to changes, they are of the form:

```
gpg: directory '/root/.gnupg' created
gpg: keybox '/root/.gnupg/pubring.kbx' created
gpg: keyserver receive failed: Cannot assign requested address
```

**Comment by ndellingwood** [comment]
retest this please

File: cmake/kokkos_tribits.cmake [hunk 1]
```diff
@@ -207,6 +207,11 @@ MACRO(KOKKOS_SETUP_BUILD_ENVIRONMENT)
     IF (NOT KOKKOS_HAS_TRILINOS)
       SET(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${Kokkos_SOURCE_DIR}/cmake/Modules/")
       INCLUDE(${KOKKOS_SRC_PATH}/cmake/kokkos_tpls.cmake)
+    ELSE()
+      IF (Kokkos_ENABLE_LIBDL)
+        KOKKOS_IMPORT_TPL(LIBDL)
+      ENDIF()
+    ENDIF()
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake:214]:
```suggestion
```

Comment (ndellingwood) [reply cmake/kokkos_tribits.cmake:214]: Force-pushed fix, must've typoed when cleaning up, thanks for catch

**Comment by jjwilke** [comment]
What is the problem when it doesn't get set?  We've never enabled any TPLs with Tribits.  What compile error are people getting?

**Comment by tasmith4** [comment]
@jjwilke no compile error, but the Kokkos tools don't get picked up when setting `KOKKOS_PROFILE_LIBRARY`

**Comment by ndellingwood** [comment]
@jjwilke also manually setting `-DKokkos_ENABLE_LIBDL` does not result in the macro `KOKKOS_ENABLE_LIBDL` being defined.

**Comment by jjwilke** [comment]
Is it just the macro? We should also be missing the `-ldl` on the link.

**Comment by ndellingwood** [comment]
@jjwilke I looked back at my Trilinos output and the change here got passed a warning that Kokkos_ENABLE_LIBDL is unused, but it still does not define the macro in the config.h file; -ldl may be problematic as well

**Comment by jjwilke** [comment]
I would first try to fix this in Trilinos and then we'll push back here. Have a look here: 

https://github.com/trilinos/Trilinos/pull/8108

Hopefully this resolves the issues

**Comment by dalg24** [comment]
@jjwilke @ndellingwood Any update on this?

**Comment by ndellingwood** [comment]
@dalg24 closing this as it did not resolve the Trilinos issue, we'll need a new Kokkos PR matching trilinos/Trilinos#8108 once that makes it into Trilinos; there are some TriBITS updates in progress that are necessary to support trilinos/Trilinos#8108

**Pull request closed without merging**